Enable changing times on volume attach/detach reconciling sync to fixing impact to AWS

chrislovecnm opened the pull request:
#**What this PR does / why we need it**:

We are currently blocked by API timeouts with PV volumes.  See https://github.com/kubernetes/kubernetes/issues/39526.  This is a workaround, not a fix.

**Special notes for your reviewer**:

A second PR will be dropped with CLI cobra options in it, but we are starting with increasing the reconciliation periods.  I am dropping this without major testing and will test on our AWS account. Will be marked WIP until I run smoke tests.

**Release note**:

```release-note
Provide kubernetes-controller-manager flags to control volume attach/detach reconciler sync.  The duration of the syncs can be controlled, and the syncs can be shut off as well. 
```

chrislovecnm committed:
The capability to control duration via controller-manager flags,
and the option to shut off reconciliation.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39551)
<!-- Reviewable:end -->

patrickmcclory reviewed (approved):
this doesn't actually fix the underlying problem but will unblock users who are seeing AWS API call flooding due to volume checks.

saad-ali commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -46,8 +46,10 @@ import (
+	// successive executions. This has been increased from every 100 ms to
+	// 1 minute since the timing has created an enormous amount of API traffic on
+	// such clouds as AWS.
+	reconcilerLoopPeriod time.Duration = 1 * time.Minute
```
This doesn't increase or decrease the rate of API calls, and break a lot of things (by reducing the rate at which the attach detach controller can react, so let's leave this alone.

chrislovecnm replied:
will do

saad-ali commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -61,7 +63,9 @@ const (
-	reconcilerSyncDuration time.Duration = 5 * time.Second
+	// This has been increased from every 5 seconds to every 5 minutes since
+	// the timing has created an enormous amount of API traffic on such clouds as AWS.
+	reconcilerSyncDuration time.Duration = 5 * time.Minute
```
Let's make this 1 minute instead of 5 (which will still reduce the rate 12x times). There is a tradeoff--without this check we could potentially reintroduce the AWS bug of wrong volume being attached.

Also maybe make this value configurable/enabled/disable in the same PR?

chrislovecnm replied:
@saad-ali that is the second PR that I wanted to drop.  Since I am not super familiar with wiring in the cobra options.

saad-ali reviewed (commented):
Couple of early comments

chrislovecnm commented:
Will let this build run, and then wire in the cobra options.  @saad-ali has asked for both in the same PR.

chrislovecnm commented:
A couple of unit tests are failing on osx.  I am going to run some build tests tonight.

saad-ali commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -80,6 +80,8 @@ func (s *CloudControllerManagerServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Int32Var(&s.KubeAPIBurst, "kube-api-burst", s.KubeAPIBurst, "Burst to use while talking with kubernetes apiserver")
 	fs.DurationVar(&s.ControllerStartInterval.Duration, "controller-start-interval", s.ControllerStartInterval.Duration, "Interval between starting controller managers.")
+	fs.BoolVar(&s.DisableReconciliation, "disable-reconcile", false, "Disable Volume Reconcilation")
+	fs.DurationVar(&s.ReconcilerLoopPeriod, "reconcile-loop-period", 5 * time.Minute,"The wait time between volume reconciliation")
```
Default 1 minutes please :)

chrislovecnm replied:
That is really short. What is the side effect of doing at 5?

gnufied replied:
I had another idea - but it may not work if `ReconcilerLoopPeriod` is declared as a Duration. We can keep one variable `int32` or something and if it is `-1` - that means disable the sync otherwise the value means sync interval in seconds.

chrislovecnm replied:
Understand that, but I alway look for the big red button. We are shutting down important code, and accidentally shutting it down would not be nice.

saad-ali commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -80,6 +80,8 @@ func (s *CloudControllerManagerServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Float32Var(&s.KubeAPIQPS, "kube-api-qps", s.KubeAPIQPS, "QPS to use while talking with kubernetes apiserver")
 	fs.Int32Var(&s.KubeAPIBurst, "kube-api-burst", s.KubeAPIBurst, "Burst to use while talking with kubernetes apiserver")
 	fs.DurationVar(&s.ControllerStartInterval.Duration, "controller-start-interval", s.ControllerStartInterval.Duration, "Interval between starting controller managers.")
+	fs.BoolVar(&s.DisableReconciliation, "disable-reconcile", false, "Disable Volume Reconcilation")
```
How about `DisableAttachDetachReconcilerSync`? So it is clear this flag is for the attach detach controller, and this is the sync subcomponent of the reconciler (not the whole reconciler). And `ReconcilerSyncLoopPeriod`?

chrislovecnm replied:
Done

saad-ali commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -77,23 +79,30 @@ type reconciler struct {
-		if time.Since(rc.timeOfLastSync) > rc.syncDuration {
+
+		if rc.disableReconciliation {
+			glog.V(5).	Info("Not reconciling volumes as reconciliation is diabled via the command line flag")
```
How about `Skipping "attached volumes still attached" check since it is disabled via the command line`.

chrislovecnm replied:
Let me know how my change strikes you :)

saad-ali commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -77,23 +79,30 @@ type reconciler struct {
+		if rc.disableReconciliation {
+			glog.V(5).	Info("Not reconciling volumes as reconciliation is diabled via the command line flag")
+		} else if time.Since(rc.timeOfLastSync) > rc.syncDuration {
+			glog.V(5).Info("Reconciling volumes")
```
How about `Starting "attached volumes still attached" check`.

chrislovecnm replied:
Will re-word, not sure if I understand what you are saying :)

chrislovecnm replied:
Let me know how my change strikes you :)

saad-ali reviewed (commented):
Some preliminary comments.

chrislovecnm commented:
And I broke the build.  Will fix on linux ... SGTM

chrislovecnm commented:
Review please @kris-nova

jingxu97 commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -77,23 +79,30 @@ type reconciler struct {
-		// are still attached to the node and udpate the status if they are not.
-		if time.Since(rc.timeOfLastSync) > rc.syncDuration {
+
+		if rc.disableReconciliation {
```
We are not disable reconciliation. DisableAttachDetachReconcilerSync flag Saad suggested is only for disable rc.sync() below. 

So rc. reconcile() should be be changed. only change as below
if !rc. DisableAttachDetachReconcilerSync {
   rc.Sync()
}

jingxu97 replied:
I guess this might be the reason for the test failures.

chrislovecnm replied:
Fixed

chrislovecnm commented:
@jingxu97 / @saad-ali am notice that https://github.com/kubernetes/kubernetes/commit/ebf796a649397b0bda8e64e0e3b46d585e0a4443#diff-2b533d4422a255f4bd8521394e285679 shows that we have openapi auto-generated.  What is the magic command for that??

liggitt commented on pkg/apis/componentconfig/types.go:
```diff
@@ -772,6 +773,14 @@ type KubeControllerManagerConfiguration struct {
+	// This flag enables or disables reconcile.  Is false by default, and thus enabled.
+	DisableReconciliation bool
+	// ReconcilerSyncDuration is the amount of time the reconciler sync states loop
+	// wait between successive executions. Is set to 5 min by default.
```
These names should make it clearer these options are related to volume attach/detach reconciliation

chrislovecnm replied:
@saad-ali made some recommendations

liggitt commented on pkg/apis/componentconfig/types.go:
```diff
@@ -772,6 +773,14 @@ type KubeControllerManagerConfiguration struct {
+	// Reconciler runs a periodic loop to reconcile the desired state of the with
+	// the actual state of the world by triggering attach detach operations.
+	// This flag enables or disables reconcile.  Is false by default, and thus enabled.
+	DisableReconciliation bool
```
What are the side effects of disabling reconciliation? Is reconciliation the mechanism that updates mounted config file and secret volumes when they change in the API?

chrislovecnm replied:
@saad-ali care to comment?

liggitt commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -58,10 +58,6 @@ const (
-
-	// reconcilerSyncDuration is the amount of time the reconciler sync states loop
-	// wait between successive executions
-	reconcilerSyncDuration time.Duration = 5 * time.Second
```
Is this changing from 5 seconds to 5 minutes?

chrislovecnm replied:
Yes, and @saad-ali are discussing it.  @saad-ali is recommending 1 min, and my question is what is the implication of 5 min.  Read the referencing issue for more information why we are backing this off.

liggitt replied:
If reconciliation has varying cost for different volume types, is a single tuning interval across all types what we want?

jingxu97 replied:
make reconcilerSyncDuration time very long might cause problem discussed in issue #33760 when state is out of sync. It explained the main reason we added sync loop here. Since this PR makes configurable, I think 1 min is good enough for most of the cases. This is a tradeoff between overhead and reducing the window of state out sync.

chrislovecnm replied:
What would you recommend? We are planning to drop this in as an emergency release on Tuesday.

As I mentioned in https://github.com/kubernetes/kubernetes/issues/39526 

TLDR;

Anyone that is using 1.4.6 or 1.4.7 in AWS will exceed their rate limits with as little as 20 PV attached to a cluster. We have one account that is at about 24k API calls per hour because of timeouts. This makes PV unusable on AWS.

chrislovecnm replied:
@jingxu97 will update the default to 1 min, per your and @saad-ali recommendation.  Will wait on e2e to finish :)  Really appreciate everyone help with this!!

jingxu97 replied:
The number of PV should not matter since the syncState() issues requests per node not per volume. If your AWS cluster has a large number of nodes or has multiple zones of clusters, it will cause the problem. Since this PR makes it configurable, user can set it up depending on their cluster setup. What is the API call rate limit? 

Suppose it is set to 1 minute, since there are two API calls in for each node, the call rate will be Nx2x60 per hour

liggitt replied:
Can you summarize what the impact of lengthening this sync interval is for volumes in general? The referenced issue talks about the downside  of synchronizing AWS volumes frequently, but I assume there's also a downside of synchronizing less frequently (or never). Does that impact pod startup time? Termination time? Correction of volume mount issues? Need to document that somewhere if this is tunable and we want cluster admins to understand the trade offs. It also seems like tuning a single sync period for all volumes types will be difficult if some (like AWS) are relatively expensive and some (like hostPath or NFS) or relatively cheap

chrislovecnm replied:
Can you open an issue?  I agree that we need to discuss more.

liggitt replied:
Can you briefly describe the impact here before we cherry pick this change into old releases?

chrislovecnm replied:
@saad-ali I will let you run with this conversation.

chrislovecnm replied:
@gnufied care to comment?

jingxu97 replied:
The impact to change the default sync state period to 1 minute is that it increases the chance of hitting the issue described in #33760 (the wrong volume might be mounted to the pod because the volume attach/detach state is out of sync in the local cache of reconciler running on master node). Please refer to #33760 the section of "Will the proposed approach solve the problem completely?".

smarterclayton replied:
I agree with Jordan, as a concerned lead I have no idea what the impacts or risks to this change are, and all I'm seeing is "might", "could", and "flake" on both sides of the fix / existing state.

Anything that needs to get cherry-picked that is important had better have a clear definition so people can understand them.  

Please add that - removing the label until someone can summarize.

smarterclayton replied:
Specifically, a comment was made:

> this increases the change of mounting the wrong volume to the pod

which is incredibly terrifying.

gnufied replied:
The wrong volume will be only mounted if:

1. Volume is detached from pod, outside of k8s system (like AWS console)
2. Node is rebooted. 

#1 is not a huge concern because - usually we don't recommend users to detach volumes attached to nodes/pods. 

#2 is bit of a problem and I am damned for saying this - but as long as time it takes for a reboot to complete is more than time specified here - we would be okay.

smarterclayton replied:
Was that an "OR" or an "AND"?  If it's an AND, that resolves my concern (because I agree 1 is PIBKAC).  If it's an OR, then that's a really important thing to document, and very scary, and should be part of the flag documentation.

jingxu97 replied:
It is an "OR". @gnufied lists two possible scenarios that the problem might happen.

chrislovecnm reviewed (commented):
Comments

chrislovecnm commented:
@k8s-bot unit test this

chrislovecnm commented:
@saad-ali / @jingxu97 I am leaving this in WIP, until I run some base smoke test tomorrow.  Will update the issue when I would like a final review.

jingxu97 commented:
reconciliation is main loop for controller to attach/detach volumes. It
checks the current state and desired state and determine whether to attach
or detach volumes. If it is disabled, attach/detach will not happen at all.

On Fri, Jan 6, 2017 at 9:34 PM, Jordan Liggitt <notifications@github.com>
wrote:

> *@liggitt* commented on this pull request.
> ------------------------------
>
> In pkg/apis/componentconfig/types.go
> <https://github.com/kubernetes/kubernetes/pull/39551#pullrequestreview-15599952>
> :
>
> > @@ -772,6 +773,14 @@ type KubeControllerManagerConfiguration struct {
>  	// Zone is treated as unhealthy in nodeEvictionRate and secondaryNodeEvictionRate when at least
>  	// unhealthyZoneThreshold (no less than 3) of Nodes in the zone are NotReady
>  	UnhealthyZoneThreshold float32
> +	// Reconciler runs a periodic loop to reconcile the desired state of the with
> +	// the actual state of the world by triggering attach detach operations.
> +	// This flag enables or disables reconcile.  Is false by default, and thus enabled.
> +	DisableReconciliation bool
>
> What are the side effects of disabling reconciliation? Is reconciliation
> the mechanism that updates mounted config file and secret volumes when they
> change in the API?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/39551#pullrequestreview-15599952>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ASSNxRB5Ta6aFICgmc1QQgOjEKoIXYoDks5rPyPZgaJpZM4LdMaG>
> .
>



-- 
- Jing

chrislovecnm commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -80,6 +80,8 @@ func (s *CloudControllerManagerServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Int32Var(&s.KubeAPIBurst, "kube-api-burst", s.KubeAPIBurst, "Burst to use while talking with kubernetes apiserver")
 	fs.DurationVar(&s.ControllerStartInterval.Duration, "controller-start-interval", s.ControllerStartInterval.Duration, "Interval between starting controller managers.")
+	fs.BoolVar(&s.DisableAttachDetachReconcilerSync, "disable-attach-detach-reconcile", false, "Disable volume attach detach reconcilation")
+	fs.DurationVar(&s.ReconcilerSyncLoopPeriod, "reconcile-sync-loop-period", 5*time.Minute, "The wait time between volume attach detach reconciliation")
```
@saad-ali / @jingxu97 & et al, I am noticing that other durations are like `ControllerStartInterval metav1.Duration`.  Does the `ReconcilerSyncLoopPeriod` default need to be this as well?  If so how the heck do I set the default value?  

Ah API machinery code ....  Feel like such a n00b ... love it!

jingxu97 replied:
I think you might set it up as line 47
NodeMonitorPeriod:       metav1.Duration{Duration: 5 * time.Second},

liggitt replied:
Yes, follow the patten of the existing vars. Pass the current value in as the default for the flag. Define the default you want in NewCloudControllerManagerServer

chrislovecnm commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/controller/volume/attachdetach/attach_detach_controller.go, line 68 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_pwpEWWZWUfALIsnGJ-r1-68:-K_rbfgqILYJSKTbxpnv:bst5qx) ([raw file](https://github.com/kubernetes/kubernetes/blob/2c7e9cf1eed0a8e27d61f62ebd8a637e81429b0f/pkg/controller/volume/attachdetach/attach_detach_controller.go#L68)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

@saad-ali that is the second PR that I wanted to drop.  Since I am not super familiar with wiring in the cobra options.
</blockquote></details>

Updated PR with cobra comments.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551)*
<!-- Sent from Reviewable.io -->

chrislovecnm commented:
Thanks!

---

Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-:-K_rbngOyBh0naCYH2lD:b-ucfknh)*
<!-- Sent from Reviewable.io -->

chrislovecnm commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/apis/componentconfig/types.go, line 779 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_qNiRENKMa8c0C0ciU-r3-779:-K_rc87u9NUn8Xmh9gsZ:bcxwr3h) ([raw file](https://github.com/kubernetes/kubernetes/blob/3aab5d2484b082ba2ad6cdbc5efe4b707d3189c9/pkg/apis/componentconfig/types.go#L779)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

@saad-ali care to comment?
</blockquote></details>

Jin responded to this.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551)*
<!-- Sent from Reviewable.io -->

chrislovecnm commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[cmd/cloud-controller-manager/app/options/options.go, line 84 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_qNiRENKMa8c0C0ciS-r4-84:-K_rgotnN4jIXvw28LDf:b-yrmq7g) ([raw file](https://github.com/kubernetes/kubernetes/blob/620477e76c5e8a9f1a2f706c2ffe06e053563b47/cmd/cloud-controller-manager/app/options/options.go#L84)):*
<details><summary><i>Previously, liggitt (Jordan Liggitt) wrote…</i></summary><blockquote>

Yes, follow the patten of the existing vars. Pass the current value in as the default for the flag. Define the default you want in NewCloudControllerManagerServer
</blockquote></details>

Fixed. Committing after e2e.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551)*
<!-- Sent from Reviewable.io -->

kris-nova commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -80,6 +81,8 @@ func (s *CloudControllerManagerServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Int32Var(&s.KubeAPIBurst, "kube-api-burst", s.KubeAPIBurst, "Burst to use while talking with kubernetes apiserver")
 	fs.DurationVar(&s.ControllerStartInterval.Duration, "controller-start-interval", s.ControllerStartInterval.Duration, "Interval between starting controller managers.")
+	fs.BoolVar(&s.DisableAttachDetachReconcilerSync, "disable-attach-detach-reconcile", false, "Disable volume attach detach reconcilation")
+	fs.DurationVar(&s.ReconcilerSyncLoopPeriod.Duration, "reconcile-sync-loop-period", s.ReconcilerSyncLoopPeriod.Duration, "The wait time between volume attach detach reconciliation")
```
Suggested to add a unit of measurement here 

```
fs.DurationVar(&s.ReconcilerSyncLoopPeriod.Duration, "reconcile-sync-loop-period", s.ReconcilerSyncLoopPeriod.Duration, "The wait time in minutes between volume attach detach reconciliation")
```

chrislovecnm replied:
It is any duration, should we parse only minutes?  Could be seconds like it is now.

liggitt replied:
Make it a duration, which you set with a unit (e.g. "1m", etc)

jingxu97 commented:
I think it would be better to make it second. Thanks
Chris Love <notifications@github.com>于2017年1月7日 周六下午12:13写道：

> *@chrislovecnm* commented on this pull request.
>
>
>
>
> ------------------------------
>
>
>
>
> In cmd/cloud-controller-manager/app/options/options.go
> <https://github.com/kubernetes/kubernetes/pull/39551>:
>
>
> > @@ -80,6 +81,8 @@ func (s *CloudControllerManagerServer) AddFlags(fs *pflag.FlagSet) {
>
>  	fs.Float32Var(&s.KubeAPIQPS, "kube-api-qps", s.KubeAPIQPS, "QPS to use while talking with kubernetes apiserver")
>
>  	fs.Int32Var(&s.KubeAPIBurst, "kube-api-burst", s.KubeAPIBurst, "Burst to use while talking with kubernetes apiserver")
>
>  	fs.DurationVar(&s.ControllerStartInterval.Duration, "controller-start-interval", s.ControllerStartInterval.Duration, "Interval between starting controller managers.")
>
> +	fs.BoolVar(&s.DisableAttachDetachReconcilerSync, "disable-attach-detach-reconcile", false, "Disable volume attach detach reconcilation")
>
> +	fs.DurationVar(&s.ReconcilerSyncLoopPeriod.Duration, "reconcile-sync-loop-period", s.ReconcilerSyncLoopPeriod.Duration, "The wait time between volume attach detach reconciliation")
>
>
>
> It is any duration, should we parse only minutes? Could be seconds like it
> is now.
>
>
>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/39551>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ASSNxZ6RZlQ1xDrnKLaoJcmhGskqSZWtks5rP_HygaJpZM4LdMaG>
> .
>

chrislovecnm commented:
@k8s-bot unit test this

chrislovecnm commented:
@jingxu97 / @saad-ali & et al - aws e2e is failing on PetSets/StatefulSets, which is a bit concerning.  Trying to get a test running.

I do need some help figuring out what the unit/integration tests are not happy.  They are passing locally.

chrislovecnm commented:
@saad-ali 1 min still seems a bit high.  Can we look at 5 min?  What is the consequence of 5 min?

chrislovecnm commented:
Review status: 0 of 9 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/controller/volume/attachdetach/reconciler/reconciler.go, line 101 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_qNiRENKMa8c0C0ciW-r2-101:-K_zAmZXrwSzQC7ROlfi:bn0ut40) ([raw file](https://github.com/kubernetes/kubernetes/blob/a5ea3757085a0123645fa0a41fb3270f4641cfb8/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L101)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

Fixed
</blockquote></details>

Actually, I had a bug, thank goodness for AWS e2e, it caught that I updated the wrong cobra options file and the cloud-controller-manager.   Now it is passing!  AWS e2e for the win.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551)*
<!-- Sent from Reviewable.io -->

chrislovecnm commented:
Review status: 0 of 9 files reviewed at latest revision, 11 unresolved discussions.

---

*[cmd/cloud-controller-manager/app/options/options.go, line 84 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_qNiRENKMa8c0C0ciS-r2-84:-K_zB5Gi7Vnt-eBr476U:b-er8ebp) ([raw file](https://github.com/kubernetes/kubernetes/blob/a5ea3757085a0123645fa0a41fb3270f4641cfb8/cmd/cloud-controller-manager/app/options/options.go#L84)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

Understand that, but I alway look for the big red button. We are shutting down important code, and accidentally shutting it down would not be nice.
</blockquote></details>

You ok with this?  What are your thoughts?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551)*
<!-- Sent from Reviewable.io -->

chrislovecnm commented:
Review status: 0 of 9 files reviewed at latest revision, 11 unresolved discussions.

---

*[cmd/cloud-controller-manager/app/options/options.go, line 85 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_qNiRENKMa8c0C0ciS-r5-85:-K_zBE8ByYBbIWS-h9G2:b-itg4wt) ([raw file](https://github.com/kubernetes/kubernetes/blob/997f040439cc85530008e8b5b680a5862bccb460/cmd/cloud-controller-manager/app/options/options.go#L85)):*
<details><summary><i>Previously, liggitt (Jordan Liggitt) wrote…</i></summary><blockquote>

Make it a duration, which you set with a unit (e.g. "1m", etc)
</blockquote></details>

Done, and we have a check that it cannot be < 1 second.  Which is insane small amount of time, but we are checking for it so that we don't create a tight loop.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551)*
<!-- Sent from Reviewable.io -->

chrislovecnm commented:
Reviewed 4 of 8 files at r4, 1 of 3 files at r5, 1 of 1 files at r6, 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551)*
<!-- Sent from Reviewable.io -->

chrislovecnm commented:
@saad-ali this is no longer WIP. We need to discuss 1 vs 5 min, before we merge.  I have tested on a cluster  we 70 pvc mounted. I need to cherry pick this myself into 1.4.x stream and run more tests. 1.6 is running like a champ.

saad-ali commented:
> @saad-ali this is no longer WIP. We need to discuss 1 vs 5 min, before we merge. I have tested on a cluster we 70 pvc mounted. I need to cherry pick this myself into 1.4.x stream and run more tests. 1.6 is running like a champ.

Taking a look

saad-ali commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -77,23 +79,33 @@ type reconciler struct {
+		if rc.disableReconciliation {
+			glog.V(5).Info("Skipping reconciling attached volumes still attached since it is disabled via the command line.")
+		} else if rc.syncDuration < time.Second {
+			glog.V(5).Info("Skipping reconciling attached volumes still attached since it is set to less than one secdon via the command line.")
```
secdon -> second

saad-ali commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -86,15 +86,16 @@ func NewCMServer() *CMServer {
+			ConcurrentGCSyncs:        20,
+			ClusterSigningCertFile:   "/etc/kubernetes/ca/ca.pem",
+			ClusterSigningKeyFile:    "/etc/kubernetes/ca/ca.key",
+			ReconcilerSyncLoopPeriod: metav1.Duration{Duration: 1 * time.Minute},
```
1 minute is good. And since the value is now configurable, you can always tweak it in your deployments, if 1 min is not sufficient.

chrislovecnm replied:
Going to do some load testing this evening. Working great at 70 pv, want to push to 2-3 hundred.

saad-ali replied:
Sweet.

saad-ali commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -181,6 +182,8 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Int32Var(&s.LargeClusterSizeThreshold, "large-cluster-size-threshold", 50, "Number of nodes from which NodeController treats the cluster as large for the eviction logic purposes. --secondary-node-eviction-rate is implicitly overridden to 0 for clusters this size or smaller.")
 	fs.Float32Var(&s.UnhealthyZoneThreshold, "unhealthy-zone-threshold", 0.55, "Fraction of Nodes in a zone which needs to be not Ready (minimum 3) for zone to be treated as unhealthy. ")
+	fs.BoolVar(&s.DisableAttachDetachReconcilerSync, "disable-attach-detach-reconcile", false, "Disable volume attach detach reconcilation")
+	fs.DurationVar(&s.ReconcilerSyncLoopPeriod.Duration, "reconcile-sync-loop-period", s.ReconcilerSyncLoopPeriod.Duration, "The wait time between volume attach detach reconciliation")
```
Mention attach detach in the name and description so it is clear that this is used by the attach/detach controller.

chrislovecnm replied:
What do you mean?  I updated `DisableAttachDetachReconcilerSync`.   You want me to refactor the other as well?

saad-ali replied:
Right. It's not obvious that `ReconcilerSyncLoopPeriod` is reconciler for which controller. But let's just leave it as is. I'd like to get this merged quickly so it has time to bake before the release.

saad-ali commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -181,6 +182,8 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Float32Var(&s.SecondaryNodeEvictionRate, "secondary-node-eviction-rate", 0.01, "Number of nodes per second on which pods are deleted in case of node failure when a zone is unhealthy (see --unhealthy-zone-threshold for definition of healthy/unhealthy). Zone refers to entire cluster in non-multizone clusters. This value is implicitly overridden to 0 if the cluster size is smaller than --large-c ...
 	fs.Int32Var(&s.LargeClusterSizeThreshold, "large-cluster-size-threshold", 50, "Number of nodes from which NodeController treats the cluster as large for the eviction logic purposes. --secondary-node-eviction-rate is implicitly overridden to 0 for clusters this size or smaller.")
 	fs.Float32Var(&s.UnhealthyZoneThreshold, "unhealthy-zone-threshold", 0.55, "Fraction of Nodes in a zone which needs to be not Ready (minimum 3) for zone to be treated as unhealthy. ")
+	fs.BoolVar(&s.DisableAttachDetachReconcilerSync, "disable-attach-detach-reconcile", false, "Disable volume attach detach reconcilation")
```
"reconcilation" -> "reconciler sync" both here and below to make it clear that this is not disabling or modifying the reconciler rate, but a sub-function of the reconciler which is the sync

chrislovecnm replied:
Changed

saad-ali reviewed (changes requested):
A few minor comments. Let's get those addressed, and get this merged and cherrypicked by EOD

chrislovecnm commented:
Great comment, can't we open an issue?

---

Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[cmd/kube-controller-manager/app/options/options.go, line 186 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_yqy-DVLFgiIme44WU-r7-186:-Ka-PeLR3fHzZSh1Lv1G:b18popr) ([raw file](https://github.com/kubernetes/kubernetes/blob/29435e6312bab8ee0050915c4ba66bad24f26f01/cmd/kube-controller-manager/app/options/options.go#L186)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

Mention attach detach in the name and description so it is clear that this is used by the attach/detach controller.
</blockquote></details>

Will change when I get home

---

*[pkg/controller/volume/attachdetach/reconciler/reconciler.go, line 106 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_qNiRENKMa8c0C0ciW-r7-106:-Ka-Pmo1RAFmzqYlQzhh:b-16z66u) ([raw file](https://github.com/kubernetes/kubernetes/blob/29435e6312bab8ee0050915c4ba66bad24f26f01/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L106)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

secdon -> second
</blockquote></details>

Will update, noticed earilier, but wanted to wait on e2e.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-:-Ka-PMOtRiz40EoQV3tu:bugc45p)*
<!-- Sent from Reviewable.io -->

chrislovecnm commented:
@saad-ali do you mind waiting for the merge and cherry pick till I get a 1.4.x cluster up myself? Will be later tonight.

saad-ali commented:
> @saad-ali do you mind waiting for the merge and cherry pick till I get a 1.4.x cluster up myself? Will be later tonight.

No problem

chrislovecnm commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[cmd/kube-controller-manager/app/options/options.go, line 186 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_yqy-DVLFgiIme44WU-r7-186:-Ka-xKjhTp0HwCaufxhD:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/29435e6312bab8ee0050915c4ba66bad24f26f01/cmd/kube-controller-manager/app/options/options.go#L186)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

Will change when I get home
</blockquote></details>

Done.

---

*[pkg/controller/volume/attachdetach/attach_detach_controller.go, line 52 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_pwpEWWZWUfALIsnGJ-r1-52:-Ka-xTai03T0wy3ZLtHI:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/2c7e9cf1eed0a8e27d61f62ebd8a637e81429b0f/pkg/controller/volume/attachdetach/attach_detach_controller.go#L52)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

will do
</blockquote></details>

Done.

---

*[pkg/controller/volume/attachdetach/attach_detach_controller.go, line 68 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_pwpEWWZWUfALIsnGJ-r1-68:-Ka-xS5hDd8mXkiQt3hU:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/2c7e9cf1eed0a8e27d61f62ebd8a637e81429b0f/pkg/controller/volume/attachdetach/attach_detach_controller.go#L68)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

Updated PR with cobra comments.
</blockquote></details>

Done.

---

*[pkg/controller/volume/attachdetach/reconciler/reconciler.go, line 102 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_qNiRENKMa8c0C0ciW-r2-102:-Ka-xPLiOQ8-oL8wgbHb:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/a5ea3757085a0123645fa0a41fb3270f4641cfb8/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L102)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

Let me know how my change strikes you :)
</blockquote></details>

Done.

---

*[pkg/controller/volume/attachdetach/reconciler/reconciler.go, line 104 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_qNiRENKMa8c0C0ciW-r2-104:-Ka-xQ_MWqsZz_PM23ti:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/a5ea3757085a0123645fa0a41fb3270f4641cfb8/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L104)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

Let me know how my change strikes you :)
</blockquote></details>

Done.

---

*[pkg/controller/volume/attachdetach/reconciler/reconciler.go, line 106 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_qNiRENKMa8c0C0ciW-r7-106:-Ka-xc5LMbWqrgSSwCe6:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/29435e6312bab8ee0050915c4ba66bad24f26f01/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L106)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

Will update, noticed earilier, but wanted to wait on e2e.
</blockquote></details>

Done.

---

*[cmd/cloud-controller-manager/app/options/options.go, line 83 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_qNiRENKMa8c0C0ciS-r2-83:-Ka-xfOH2ZPZbepOiyt_:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/a5ea3757085a0123645fa0a41fb3270f4641cfb8/cmd/cloud-controller-manager/app/options/options.go#L83)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Done.

---

*[cmd/cloud-controller-manager/app/options/options.go, line 84 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_qNiRENKMa8c0C0ciS-r4-84:-Ka-xgjE3K3Q86KIjcsD:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/620477e76c5e8a9f1a2f706c2ffe06e053563b47/cmd/cloud-controller-manager/app/options/options.go#L84)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

Fixed. Committing after e2e.
</blockquote></details>

Done.

---

*[cmd/cloud-controller-manager/app/options/options.go, line 85 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551#-K_qNiRENKMa8c0C0ciS-r5-85:-Ka-xhU4aEVLxpdYRXbv:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/997f040439cc85530008e8b5b680a5862bccb460/cmd/cloud-controller-manager/app/options/options.go#L85)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary><blockquote>

Done, and we have a check that it cannot be < 1 second.  Which is insane small amount of time, but we are checking for it so that we don't create a tight loop.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39551)*
<!-- Sent from Reviewable.io -->

chrislovecnm commented:
@saad-ali so after much banging my head against the wall with a cherry pick, I have noticed that https://github.com/kubernetes/kubernetes/tree/v1.4.8-beta.0 does not compile.  You want me to file another issue?  Specifically:

```
test/e2e_node/restart_test.go:77: undefined: newTestPods
test/e2e_node/restart_test.go:79: undefined: createBatchPodWithRateControl
test/e2e_node/restart_test.go:80: undefined: deletePodsSync
```

I think I can still do a make quick release and test, and will be able to test in the AM.

saad-ali commented:
/lgtm

saad-ali commented:
> @saad-ali so after much banging my head against the wall with a cherry pick, I have noticed that https://github.com/kubernetes/kubernetes/tree/v1.4.8-beta.0 does not compile. You want me to file another issue? Specifically:

@jessfraz is the 1.4 branch broken?

gnufied commented:
> @jessfraz is the 1.4 branch broken?

1.4 branch builds fine. But this change is not a easy cherry-pick to 1.4, so I am putting it down to problematic merge.

chrislovecnm commented:
> 1.4 branch builds fine. But this change is not a easy cherry-pick to 1.4, so I am putting it down to problematic merge.

Yup.  Merging it by hand now.

saad-ali commented:
@smarterclayton @chrislovecnm @justinsb @jingxu97

Some options to address Clayton's concerns:
* Option 1: We leave the sync rate at 5 seconds, and we add the flag to allow people to customize it for their own deployments. They can increase it at their own risk.
  * Pro: Current default deployment remains in a "safe" state.
  * Con: Any cluster of significant enough size hits quota limit.
* Option 2: We set the default to 1 minute, with the big comment on the flag and release note saying that this may be more likely to result undesirable behavior (wrong PV being attached).
  * Pro: Reduce likelihood of hitting AWS quota limits.
  * Con: Increase likelihood of wrong PV (EBS) getting attached in AWS.
* Option 3: disable the attach detach controller by default for for AWS deployments, so that kubelet manages its own attach/detach.
  * Pro: no sync calls so no worries about hitting quota issues. No controller, so no worries about attaching wrong volume.
  * Con: could strand a volume if kubelet crashes (no third party to force detach the volume)

Thoughts?

Let's get a decision on this ASAP--reminder that 1.4.8 and 1.5.2 are scheduled to be cut tomorrow.

chrislovecnm commented:
>Option 1: We leave the sync rate at 5 seconds, and we add the flag to allow people to customize it for their own deployments. They can increase it at their own risk.
> Pro: Current default deployment remains in a "safe" state.
> Con: Any cluster of significant enough size hits quota limit.

We have had multiple reports that as few as 20 PV start to cause API limit errors in AWS.  So this approach will cause anyone in AWS issues.

>Option 2: We set the default to 1 minute, with the big comment on the flag and release note saying that this may be more likely to result undesirable behavior (wrong PV being attached).
>Pro: Reduce likelihood of hitting AWS quota limits.
>Con: Increase likelihood of wrong PV (EBS) getting attached in AWS.

As @smarterclayton mentioned this is concerning, unless we could have a check that ran when a node reboots.

>Option 3: disable the attach detach controller by default for for AWS deployments, so that kubelet manages its own attach/detach.
>Pro: no sync calls so no worries about hitting quota issues. No controller, so no worries about attaching wrong volume.
>Con: could strand a volume if kubelet crashes (no third party to force detach the volume)

Do we lose any features from this approach?  What about replacing a node, how does that impact this.  Do we have to enable anything on the `kubelet` side of the house?

smarterclayton commented:
Could we add the flag, but not change the default, and have AWS use the
flag with a longer sync period with a warning?  If this is primarily AWS
problem, I'd be in favor of keeping it confined to AWS.  I don't want to
disable attach/detach controller unless we truly have to.

What will it take to fix/minimize "wrong PV being attached"?  Maybe I
missed that in the thread, apologies if so. If we have a way to do it and
we think we can do it over time, having a temporary fix for AWS is not that
bad.

On Mon, Jan 9, 2017 at 2:36 PM, Chris Love <notifications@github.com> wrote:

> Option 1: We leave the sync rate at 5 seconds, and we add the flag to
> allow people to customize it for their own deployments. They can increase
> it at their own risk.
> Pro: Current default deployment remains in a "safe" state.
> Con: Any cluster of significant enough size hits quota limit.
>
> We have had multiple reports that as few as 20 PV start to cause API limit
> errors in AWS. So this approach will cause anyone in AWS issues.
>
> Option 2: We set the default to 1 minute, with the big comment on the flag
> and release note saying that this may be more likely to result undesirable
> behavior (wrong PV being attached).
> Pro: Reduce likelihood of hitting AWS quota limits.
> Con: Increase likelihood of wrong PV (EBS) getting attached in AWS.
>
> As @smarterclayton <https://github.com/smarterclayton> mentioned this is
> concerning, unless we could have a check that ran when a node reboots.
>
> Option 3: disable the attach detach controller by default for for AWS
> deployments, so that kubelet manages its own attach/detach.
> Pro: no sync calls so no worries about hitting quota issues. No
> controller, so no worries about attaching wrong volume.
> Con: could strand a volume if kubelet crashes (no third party to force
> detach the volume)
>
> Do we lose any features from this approach? What about replacing a node,
> how does that impact this. Do we have to enable anything on the kubelet
> side of the house?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/39551#issuecomment-271385034>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p8snZwA70twlrR_v39u0lyKVoeGIks5rQoxBgaJpZM4LdMaG>
> .
>

justinsb commented:
Option 3 (disabling the attach/detach controller) is something we should not undertake lightly, and I think would require a lot of testing.  I think that's risky in a patch release.  (Unless someone has secretly been testing this code path?)

jingxu97 commented:
Option 2 for me seems better. The main concern is node reboot. Normally
when node reboots, it also takes a little bit time (from my experience
about 20-30 seconds) before kubelet starts mounting process. Considering
the reboot time and this extra time, the chance that kubelet gets the
outdated (wrong) volume attachement information is small.

smarterclayton commented:
Anything we can do to isolate this fix to the affected cloud providers
seems like a net win - if it's just AWS and no other provide is affected,
it seems reasonable.

On Mon, Jan 9, 2017 at 2:48 PM, Jing Xu <notifications@github.com> wrote:

> Option 2 for me seems better. The main concern is node reboot. Normally
> when node reboots, it also takes a little bit time (from my experience
> about 20-30 seconds) before kubelet starts mounting process. Considering
> the reboot time and this extra time, the chance that kubelet gets the
> outdated (wrong) volume attachement information is small.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/39551#issuecomment-271388046>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p56pQXaEzbb4_TecXrGfNA1f1E1pks5rQo72gaJpZM4LdMaG>
> .
>

chrislovecnm commented:
@smarterclayton I am guessing that other providers are getting hammered by these API calls as well.

2 x num.nodes per second * (how many API calls you need to get volumes)
 
is a bunch.

gnufied commented:
That 2x number of calls is a BUG imo and it affects only AWS.  `getAWSInstance()` has same details as `e.describeInstance` - but we may not have to fix that since @justinsb already opened a PR that removes all that stuff.

gnufied commented:
> Option 1: We leave the sync rate at 5 seconds, and we add the flag to allow people to customize it for > their own deployments. They can increase it at their own risk.
> Pro: Current default deployment remains in a "safe" state.
> Con: Any cluster of significant enough size hits quota limit.

I like this option a lot. 

> We have had multiple reports that as few as 20 PV start to cause API limit errors in AWS. So this approach will cause anyone in AWS issues.

That may be so, but we can point people to relevant documentation about - how to fix it (i.e how to increase default sync duration). And this workaround may not be required once https://github.com/kubernetes/kubernetes/pull/39564 lands.

saad-ali commented:
Meeting to discuss options at 1:30 PM PT
If you didn't get the invite and want to join, hangout link will be: https://plus.google.com/hangouts/_/google.com/k8s-aws-storage

liggitt commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -77,23 +79,33 @@ type reconciler struct {
 	loopPeriod                time.Duration
 	maxWaitForUnmountDuration time.Duration
 	syncDuration              time.Duration
+	enableSync                bool
```
this is unused, right?

chrislovecnm replied:
Will check

chrislovecnm replied:
Have another commit I will send in after e2e runs.

chrislovecnm commented:
@saad-ali per the meeting we are going to set the default time to 5s, and let install tools set it higher.  We are in the process of deprecating kube-up.sh support for AWS anyway, so we probably will let that lie.

Do you want me to rename the flag and variable as well

```
	fs.DurationVar(&s.ReconcilerSyncLoopPeriod.Duration, "reconcile-sync-loop-period", s.ReconcilerSyncLoopPeriod.Duration, "The reconciler sync wait time between volume attach detach.")
```

We did not have the correct name in it.

liggitt commented:
> Do you want me to rename the flag and variable as well

Yes, please, especially the CLI flag that is getting exposed. Both should include the attach/detach qualifier (maybe something like `--attach-detach-sync-period`?)

chrislovecnm commented:
> Yes, please, especially the CLI flag that is getting exposed. Both should include the attach/detach qualifier (maybe something like --attach-detach-sync-period?)

We have 
`--disable-attach-detach-reconcile`
So
``--attach-detach-reconcile-period`
Makes sense to me.  

I am not going to update the variable names as frankly, I don't want to touch this code much.  I will update the flag name, the duration, and the flag options hack file.

saad-ali commented:
>> Do you want me to rename the flag and variable as well
> Yes, please, especially the CLI flag that is getting exposed. Both should include the attach/detach qualifier (maybe something like --attach-detach-sync-period?)

+1

> We have
> `--disable-attach-detach-reconcile`
> So
> `--attach-detach-reconcile-period`
> Makes sense to me.
> 
> I am not going to update the variable names as frankly, I don't want to touch this code much. I will update the flag name, the duration, and the flag options hack file.

Should include the word `sync` in both (not just reconcile).

chrislovecnm commented on pkg/controller/volume/attachdetach/attach_detach_controller_test.go:
```diff
@@ -41,7 +41,9 @@ func Test_NewAttachDetachController_Positive(t *testing.T) {
-		nil /* plugins */)
+		nil, /* plugins */
+		false,
+		time.Minute*5)
```
Need to update this as well ...

chrislovecnm commented:
I am noticing that with a build locally 
`pkg/generated/bindata.go`
Is claiming that it needs a commit.  I assume that I have missed that file?

chrislovecnm commented:
@saad-ali I have a couple of other tweaks I will push in after this.  Fixing times on two tests, and removing dead code in reconciler.  I am letting e2e run now, but will be in shortly.

saad-ali commented:
> @saad-ali I have a couple of other tweaks I will push in after this. Fixing times on two tests, and removing dead code in reconciler. I am letting e2e run now, but will be in shortly.

Ack.

chrislovecnm commented:
@saad-ali last e2e was successful, now waiting on this one to run.  The release notes are up to date now, and the changes that were requested are in.

saad-ali commented:
Cool, I'll take another look

saad-ali commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -181,6 +182,8 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Int32Var(&s.LargeClusterSizeThreshold, "large-cluster-size-threshold", 50, "Number of nodes from which NodeController treats the cluster as large for the eviction logic purposes. --secondary-node-eviction-rate is implicitly overridden to 0 for clusters this size or smaller.")
 	fs.Float32Var(&s.UnhealthyZoneThreshold, "unhealthy-zone-threshold", 0.55, "Fraction of Nodes in a zone which needs to be not Ready (minimum 3) for zone to be treated as unhealthy. ")
+	fs.BoolVar(&s.DisableAttachDetachReconcilerSync, "disable-attach-detach-reconcile", false, "Disable volume attach detach reconciler sync. Disabling this may cause volumes to be mismatched with pods. Use wisely.")
+	fs.DurationVar(&s.ReconcilerSyncLoopPeriod.Duration, "attach-detach-reconcile-period", s.ReconcilerSyncLoopPeriod.Duration, "The reconciler sync wait time between volume attach detach. This duration must be larger than one second, and increasing this value from the default my allow for volume mismatches.")
```
`attach-detach-reconcile-period` -> `attach-detach-reconcile-sync-period`

saad-ali replied:
`default my allow for volume` -> `default my increase the likelihood of volume`?

chrislovecnm replied:
typo fixing

saad-ali commented on hack/verify-flags/known-flags.txt:
```diff
@@ -635,3 +635,5 @@ garbage-collector-enabled
 log-lines-total
 run-duration
+disable-attach-detach-reconcile
+attach-detach-reconcile-period
```
Remember to update these after rename above.

chrislovecnm replied:
Done

saad-ali commented on pkg/apis/componentconfig/types.go:
```diff
@@ -772,6 +772,13 @@ type KubeControllerManagerConfiguration struct {
+	// This flag enables or disables reconcile.  Is false by default, and thus enabled.
+	DisableAttachDetachReconcilerSync bool
+	// ReconcilerSyncLoopPeriod is the amount of time the reconciler sync states loop
+	// wait between successive executions. Is set to 5 min by default.
```
`5 min` -> `5 sec`

chrislovecnm replied:
Done

saad-ali commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -57,6 +57,7 @@ func NewReconciler(
 	loopPeriod time.Duration,
 	maxWaitForUnmountDuration time.Duration,
 	syncDuration time.Duration,
+	disableReconciliation bool,
```
`disableReconciliation`->`disableReconcilerSync`?

saad-ali replied:
Ditto for `attach_detach_controller.go`

chrislovecnm replied:
Done, and renamed a couple stragglers.

chrislovecnm commented:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/39551/pull-kubernetes-unit/12293/ <- same open API stuff that was failing over the weekend.  No idea what is up

chrislovecnm committed:
updates from review

chrislovecnm commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -408,6 +408,10 @@ func StartControllers(controllers map[string]InitFunc, s *options.CMServer, root
 	time.Sleep(wait.Jitter(s.ControllerStartInterval.Duration, ControllerStartJitter))
 
+	if s.ReconcilerSyncLoopPeriod.Duration < time.Second {
+		return fmt.Errorf("Duration time must be greater than one second as set via command line option reconcile-sync-loop-period.  One minute is recommended.")
```
Done

chrislovecnm commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -181,6 +182,8 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Float32Var(&s.SecondaryNodeEvictionRate, "secondary-node-eviction-rate", 0.01, "Number of nodes per second on which pods are deleted in case of node failure when a zone is unhealthy (see --unhealthy-zone-threshold for definition of healthy/unhealthy). Zone refers to entire cluster in non-multizone clusters. This value is implicitly overridden to 0 if the cluster size is smaller than --large-c ...
 	fs.Int32Var(&s.LargeClusterSizeThreshold, "large-cluster-size-threshold", 50, "Number of nodes from which NodeController treats the cluster as large for the eviction logic purposes. --secondary-node-eviction-rate is implicitly overridden to 0 for clusters this size or smaller.")
 	fs.Float32Var(&s.UnhealthyZoneThreshold, "unhealthy-zone-threshold", 0.55, "Fraction of Nodes in a zone which needs to be not Ready (minimum 3) for zone to be treated as unhealthy. ")
+	fs.BoolVar(&s.DisableAttachDetachReconcilerSync, "disable-attach-detach-reconcile", false, "Disable volume attach detach reconciler sync. Disabling this may cause volumes to be mismatched with pods. Use wisely.")
```
Done

saad-ali commented:
> https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/39551/pull-kubernetes-unit/12293/ <- same open API stuff that was failing over the weekend. No idea what is up

Failing UT issue looks like https://github.com/kubernetes/kubernetes/issues/39604 which is a marked a flake, so it should pass if we run it often enough.

saad-ali commented:
@k8s-bot bazel test this

saad-ali commented:
/lgtm

saad-ali commented:
Marking P0 to get merged ASAP for tomorrow's release.

saad-ali commented:
Chris this is merged. Can you please prepare the cherry pick to 1.5 and once that is complete we can do the 1.4 cherry pick
